hail/python/hailtop/utils/utils.py

```diff
@@ -56,6 +56,27 @@ def __init__(self, parallelism):
         await self._done.wait()
 
 
+class AsyncWorkerPool:
+    def __init__(self, parallelism, queue_size=1000):
+        self._queue = asyncio.Queue(maxsize=queue_size)
+
+        for _ in range(parallelism):
+            asyncio.ensure_future(self._worker())
+
+    async def _worker(self):
+        while True:
+            f, args, kwargs = await self._queue.get()
+            try:
+                await f(*args, **kwargs)
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
+            except Exception:  # pylint: disable=broad-except
+                log.exception(f'worker pool caught exception')
```

Comment:
Why do these two classes treat exceptions differently?

Reply:
One is used in the batch aioclient to submit multiple batches in parallel while not submitting more than some threshold. Equivalent of `asyncio.gather(parallelism=N)`. If we don't raise errors, then we ignore errors that should get propagated to the user.

However, the other use case is in the Batch2 scheduler to execute a bunch of work in a pool with N-parallelism. We don't want an exception to cause the pool to die and thus want to ignore the errors while still logging them. 

I tried using a flag that said whether to raise an exception or not and reusing the same class for both, but that seemed more complicated to me.

Reply:
The exceptions raised by the `AsyncThrottledGather` don't go anywhere because they happen in a future that we never check (line 53).

Reply:
We'll probably still get log messages from asyncio though. It would appear these two classes don't actually differ in their exception handling behavior.

Reply:
I tried a different approach. I modified the asyncio.gather implementation. https://github.com/python/asyncio/blob/a791d884b9110b59bada18c2dddd399e6adc40fd/asyncio/tasks.py#L602